### PR TITLE
Add ERC: Bonding curve-embedded liquidity for NFTs

### PR DESCRIPTION
**Pull request by vgorin**
Standard for embedding liquidity into ERC-721 Non-Fungible Tokens (NFTs) without modifications to ERC-721.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: ERCS/erc-draft_embedded_liquidity.md [hunk 1]
```diff
@@ -0,0 +1,404 @@
+---
```
Comment (abcoathup) [inline ERCS/erc-draft_embedded_liquidity.md:1]:
```suggestion
---
eip: 7649
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500.

Please also update the filename.

File: ERCS/erc-draft_embedded_liquidity.md [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+title: Bonding Curve-Embedded Liquidity for ERC-721 Non-Fungible Tokens (NFTs)
+description: Standard for embedding liquidity into ERC-721 Non-Fungible Tokens (NFTs) without modifications to ERC-721.
+author: AI Protocol <gm@aiprotocol.info>
+discussions-to: https://ethereum-magicians.org/t/discussion-for-bonding-curve-embedded-liquidity-for-erc-721-non-fungible-tokens-nfts/19079
```
Comment (abcoathup) [inline ERCS/erc-draft_embedded_liquidity.md:5]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7649-bonding-curve-embedded-liquidity-for-erc-721-non-fungible-tokens-nfts/19079
```

File: ERCS/erc-draft_embedded_liquidity.md [hunk 3]
```diff
@@ -0,0 +1,404 @@
+---
+title: Bonding Curve-Embedded Liquidity for ERC-721 Non-Fungible Tokens (NFTs)
```
Comment (abcoathup) [inline ERCS/erc-draft_embedded_liquidity.md:2]:
```suggestion
title: Bonding curve-embedded liquidity for NFTs
```
Consider a shorter title

File: ERCS/erc-7649.md [hunk 4]
```diff
@@ -0,0 +1,410 @@
+---
+eip: 7649
+title: Bonding curve-embedded liquidity for NFTs
+description: Embedding liquidity into Non-Fungible Tokens (NFTs) without modifications to ERC-721.
+author: AI Protocol (@ai-protocol-dev), AI Protocol <gm@aiprotocol.info>
```
Comment (gaudren) [inline ERCS/erc-7649.md:5]: If this username and email belong to the same author, you may use the format:

```suggestion
author: AI Protocol (@ai-protocol-dev) <gm@aiprotocol.info>
```

File: ERCS/erc-7649.md [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+
+## Abstract
+
+<!--
+  The Abstract is a multi-sentence (short paragraph) technical summary. This should be a very terse and human-readable version of the specification section. Someone should be able to read only the abstract to get the gist of what this specification does.
+
+  TODO: Remove this comment before submitting
+-->
+
+This Ethereum Improvement Proposal (EIP) introduces a standard for embedding Bonding Curve-like liquidity into
```
Comment (gaudren) [inline ERCS/erc-7649.md:26]:
```suggestion
This proposal introduces a standard for embedding Bonding Curve-like liquidity into
```

File: ERCS/erc-7649.md [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+  With a few exceptions, external links are not allowed. If you feel that a particular resource would demonstrate a compelling case for your EIP, then save it as a printer-friendly PDF, put it in the assets folder, and link to that copy.
+
+  TODO: Remove this comment before submitting
+-->
+
+The ERC-721 standard lacks a specific mechanism for embedding bonding curve-based liquidity, limiting the creative
+possibilities for NFT-based projects. This EIP addresses the need for a standardized approach to integrate bonding curve
+contracts seamlessly into ERC-721 NFTs, allowing for diverse and innovative implementations without modifying the
+ERC-721 standard.
```
Comment (gaudren) [inline ERCS/erc-7649.md:47]: Would be nice to include some use cases.

File: ERCS/erc-7649.md [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+    - Creators may specify and receive a percentage of transaction fees generated by the embedded liquidity
+      contract during buy and sell operations.
+
+5.  Implementation Independence:
+    - The standard is implementation-agnostic, providing developers with the freedom to choose and design the most
+      suitable liquidity mechanism for their NFT projects.
+    - The embedded liquidity contract can interact with ERC-20 tokens or native ETH, depending on the implementation
+      requirements.
+
+### Bonding Curve Spec
```
Comment (gaudren) [inline ERCS/erc-7649.md:90]:
```suggestion
### `BondingCurve` Interface
```

File: ERCS/erc-7649.md [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+	 *
+	 * @param supply total shares supply
+	 * @param amount number of shares to buy/sell
+	 * @return the price of the shares (all `amount` amount)
+	 */
+	function getPrice(uint256 supply, uint256 amount) external pure returns(uint256);
+}
+```
+
+### Bonding Curve-Embedded Liquidity / Tradable Shares Spec
```
Comment (gaudren) [inline ERCS/erc-7649.md:114]:
```suggestion
### Bonding Curve-Embedded Liquidity / `TradeableShares` Interface
```

File: ERCS/erc-7649.md [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+  The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages.
+
+  The current placeholder is acceptable for a draft.
+
+  TODO: Remove this comment before submitting
+-->
+
+The proposed standard focuses on enhancing the ERC-721 standard by introducing a framework for embedding bonding
+curve-based liquidity into NFTs. This approach provides creators with a flexible and customizable tool to attract
+liquidity through bonding curve mechanisms, while ensuring creators receive guaranteed fees for their contributions.
```
Comment (gaudren) [inline ERCS/erc-7649.md:381]: This belongs in the Motivation section. Rationale should explain the technical choices made in this proposal.

**Comment by vgorin** [comment]
Hello, Andrew @abcoathup !

This looks to get stuck on @eip-review-bot  review, though all the checks have passed. Is there anything I can do to unblock the merge?

Thank you!

**Comment by vgorin** [comment]
Hello, @gaudren , thanks for taking the time to review this again.

It looks like for the merge to happen the branch needs to be updated with the latest changes from master and in the same time be reviewed/approved, which is unlikely to happen simultaneously.

Could you please suggest how this can be unblocked? Thank you!

**Comment by gaudren** [comment]
@vgorin Once an editor approves (I am not an editor), the bot will merge for you.

**Comment by vgorin** [comment]
Hello, @abcoathup , thanks again for reviewing the EIP.

I've resolved all the review points you've raised, but probably didn't ping you directly to re-review (not sure if you receive a notification automatically). Pinging here.

Thanks.

File: ERCS/erc-7649.md [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+The rationale behind the design choices for the embedded liquidity standard is deeply rooted in providing a robust and
+versatile framework for embedding Bonding Curve-like liquidity into NFTs. The following key considerations have
+influenced the technical decisions:
+
+1.  **Bonding Curve-Embedded Liquidity / Tradable Shares Contract**:
+    - **Seamless Integration**: The decision to allow an embedded bonding curve-based liquidity contract to be attached
+        to an NFT without altering the ERC-721 standard stems from the desire for seamless integration.
+        This approach ensures that NFT developers can enhance their creations with liquidity mechanisms without
+        introducing complexities or requiring modifications to the widely adopted ERC-721 standard.
```
Comment (SamWilsn) [inline ERCS/erc-7649.md:386]: This point repeats content from line ~69. Please move content explaining "why you did things" into the Rationale section, and only keep specific requirements in the Specification section.

File: ERCS/erc-7649.md [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+
+    - **Liquidity Management**: The bonding curve contract's role in managing liquidity through the bonding curve
+        mechanism is essential. This design choice facilitates a dynamic and automated pricing model based on supply
+        and demand, contributing to the overall liquidity and tradability of NFT shares.
+
+2.  **Liquidity Mechanisms**:
+    - **Versatility**: The standard's allowance for various liquidity mechanisms, including bonding curves, liquidity
+        locking with ERC-20 tokens, or native ETH liquidity pools, reflects a commitment to versatility.
+        This flexibility enables developers to tailor the liquidity mechanism based on the specific requirements of
+        their NFT projects, promoting innovation and diversity within the ecosystem.
```
Comment (SamWilsn) [inline ERCS/erc-7649.md:396]: Again, this seems to repeat content from the specification section.

File: ERCS/erc-7649.md [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+        environment.
+
+5. **Implementation Independence**:
+    - **Developer Freedom**: The standard's implementation-agnostic approach is motivated by the desire to provide
+        developers with the freedom to choose and design the most suitable liquidity mechanism for their NFT projects.
+        Whether interacting with ERC-20 tokens or native ETH, this independence ensures that developers can make
+        informed choices based on the specific requirements of their projects.
+
+The rationale behind these design choices is to create a Tradable Shares standard that is not only technically sound but
+also flexible, adaptable, and supportive of diverse and creative implementations within the ERC-721 ecosystem.
```
Comment (SamWilsn) [inline ERCS/erc-7649.md:425]: Your rationale section is very... generic. It doesn't really give information about any specifics in your proposal. Normally the rationale section should explain individual technical choices made within the proposal itself.

A poor example for your proposal would be explaining why you chose 18 decimals for the fee percentage. As the author, I'm sure you can come up with better things to explain :rofl:

**Review by SamWilsn — COMMENTED** [review]
I'm curious, did you use ChatGPT to help write this? Isn't a problem either way.

File: ERCS/erc-7649.md (hunk 12 above)
Comment (vgorin) [reply ERCS/erc-7649.md]: Updated both Specification and Rationale sections. Made Specification section more strict, removed the support of any other liquidity mechanism, except for bonding curves – focus solely on the bonding curves.

Every item is Rationale corresponds to the item in Specification and explains "why", while items in Specification explain "what".

**Pull request merged**